static/src/javascripts/projects/common/modules/commercial/track-ad.js

```diff
@@ -0,0 +1,41 @@
+define([
+    'Promise',
+    'common/utils/mediator'
+], function (
+    Promise,
+    mediator
+) {
+    var promises;
+
+    function TrackAd() {
+        promises = {};
+    }
+
+    TrackAd.prototype.track = function (id) {
+        promises[id] = promises[id] || new Promise(function (resolve, reject) {
+            var onAdLoaded = function (event) {
+                if (event.slot.getSlotElementId() === id) {
+                    unlisten();
+                    resolve(!event.isEmpty);
+                }
+            };
+
+            var onAllAdsLoaded = function () {
+                unlisten();
+                reject(new Error('Slot ' + id + ' was never loaded'));
+            };
+
+            function unlisten() {
+                mediator.off('modules:commercial:dfp:rendered', onAdLoaded);
+                mediator.off('modules:commercial:dfp:alladsrendered', onAllAdsLoaded);
+            }
+
+            mediator.on('modules:commercial:dfp:rendered', onAdLoaded);
+            mediator.on('modules:commercial:dfp:alladsrendered', onAllAdsLoaded);
+        });
+
+        return promises[id];
+    };
+
+    return new TrackAd();
```

Comment:
We've written this with a constructor, but the way all the instances share the same `promises` map implies this could just as well be written as a singleton.